Dart/src/com/jetbrains/lang/dart/ide/runner/server/vmService/frame/DartVmServiceEvaluator.java

```diff
@@ -1,38 +1,176 @@
 
 import com.intellij.openapi.editor.Document;
+import com.intellij.openapi.fileEditor.*;
 import com.intellij.openapi.project.Project;
 import com.intellij.openapi.util.TextRange;
 import com.intellij.openapi.util.text.StringUtil;
+import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.psi.PsiDocumentManager;
 import com.intellij.psi.PsiElement;
 import com.intellij.psi.PsiFile;
+import com.intellij.psi.PsiManager;
+import com.intellij.psi.util.PsiTreeUtil;
 import com.intellij.xdebugger.XSourcePosition;
 import com.intellij.xdebugger.evaluation.ExpressionInfo;
 import com.intellij.xdebugger.evaluation.XDebuggerEvaluator;
+import com.intellij.xdebugger.frame.XValue;
 import com.jetbrains.lang.dart.ide.runner.server.vmService.DartVmServiceDebugProcess;
 import com.jetbrains.lang.dart.psi.*;
-import org.dartlang.vm.service.element.Frame;
+import com.jetbrains.lang.dart.util.DartResolveUtil;
+import gnu.trove.THashSet;
+import org.dartlang.vm.service.consumer.GetObjectConsumer;
+import org.dartlang.vm.service.element.*;
 import org.jetbrains.annotations.NotNull;
 import org.jetbrains.annotations.Nullable;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class DartVmServiceEvaluator extends XDebuggerEvaluator {
   private static final Pattern ERROR_PATTERN = Pattern.compile("Error:.* line \\d+ pos \\d+: (.+)");
 
-  @NotNull private final DartVmServiceDebugProcess myDebugProcess;
-  @NotNull private final String myIsolateId;
-  @NotNull private final Frame myFrame;
+  @NotNull protected final DartVmServiceDebugProcess myDebugProcess;
 
-  public DartVmServiceEvaluator(@NotNull final DartVmServiceDebugProcess debugProcess,
-                                @NotNull final String isolateId,
-                                @NotNull final Frame vmFrame) {
+  public DartVmServiceEvaluator(@NotNull final DartVmServiceDebugProcess debugProcess) {
     myDebugProcess = debugProcess;
-    myIsolateId = isolateId;
-    myFrame = vmFrame;
+  }
+
+  @Override
+  public boolean isCodeFragmentEvaluationSupported() {
+    return true;
+  }
+
+  @Override
+  public void evaluate(@NotNull final String expression,
+                       @NotNull final XEvaluationCallback callback,
+                       @Nullable final XSourcePosition expressionPosition) {
+    final String isolateId = myDebugProcess.getCurrentIsolateId();
+    final Project project = myDebugProcess.getSession().getProject();
+    final FileEditorManager manager = FileEditorManager.getInstance(project);
+    PsiElement element = null;
+    PsiFile psiFile = null;
+    final List<VirtualFile> libraryFiles = new ArrayList<>();
+    // Turn off pausing on exceptions as it is confusing to mouse over an expression
+    // and to have that trigger pausing at an exception.
+    myDebugProcess.getVmServiceWrapper().setExceptionPauseMode(ExceptionPauseMode.None);
```

Comment:
I agree that with the current VM behavior we do need to disable pausing on exceptions during evaluation. Maybe we'd also ask VM Service guys to enable such behavior by default? VM knows when an exception happens during program execution and when - during evaluation, so it probably could always ignore exceptions on evaluation.

Reply:
Agreed. I've asked the VM Service team to provide a service protocol option to enable that behavior.